HPCC-30365 Add XRef Sasha service to K8s

jackdelv opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jakesmith commented on system/jlib/jutil.cpp:
```diff
@@ -1899,13 +1899,24 @@ void clearThreadLocal()
+                    while (slash > start && *slash != PATHSEPCHAR)
+                        slash--;
+                    buf.insert(slash-start, PATHSEPCHAR);
+                    buf.insert((slash+1)-start, p+1);
```
not safe, buf may have reallocated after 1st insert, meaning 'slash' points to a free'd pointer.

jakesmith replied:
These insert also don't look right in that they're inserting "/<partnum>", but the mask is typically something like "myfilename._$P$\_of\_10", so it's going to end up with e.g. "/5myfilename._5_of_10" afaics.

If it did find a slash, it would prob be okay, but there isn't necessarily a full path so there's no guarantee there's a slash.

jakesmith replied:
Also, the mask can be "myfilename._$P$\_of\_$N$" (see 'N' handleing below)
This insert code is going to be hit twice in that case, and therefore insert the dir-per-part directory twice...

jackdelv replied:
Reverted changes to expandMask in favor of storing prefix and scope directory paths in cFileDesc and building up full paths in getPartName instead.

jakesmith commented on system/jlib/jutil.cpp:
```diff
@@ -1899,13 +1899,24 @@ void clearThreadLocal()
+StringBuffer &expandMask(StringBuffer &buf, const char *mask, unsigned p, unsigned n, unsigned stripeNum, bool dirPerPart)
 {
+    if (stripeNum>0)
+        addPathSepChar(buf.append('d').append(stripeNum));
```
hm, how is this strip dir going to be in right place?

jackdelv replied:
Removed changes to expandMask.

jakesmith reviewed (changes requested):
@jackdelv - I think changes look logical, but there are various problems with the changes to expandMask.
Could do with some unittests added to prove for various input it is creating the expected output.

jackdelv commented on system/jlib/jutil.cpp:
```diff
@@ -1899,13 +1899,24 @@ void clearThreadLocal()
             char next = *(s++);
             if (next=='$') {
+                if (dirPerPart)
+                {
```
Removed change.

jackdelv commented on system/jlib/jutil.cpp:
```diff
@@ -1899,13 +1899,24 @@ void clearThreadLocal()
+                {
+                    const char * start = buf.str();
+                    const char * slash = start + buf.length();
+                    while (slash > start && *slash != PATHSEPCHAR)
```
Reverted changed.

jackdelv commented on dali/sasha/saxref.cpp:
```diff
@@ -934,9 +936,22 @@ class CNewXRefManager: public CNewXRefManagerBase
+                        isDirPerPart = false;
+                }
+                if (isDirPerPart) {
+                    // MORE: Should maybe check this doesn't contain any subdirectories to make
```
Added check (pdir->dirs.ordinality()>0)

jakesmith reviewed (changes requested):
As discussed in our meeting, there's are some issues with the way the info. is being scanned and stored in the hierarchical structure built up during the physical file scan and used subsequently.
Currently too much information is being stored per cFileDesc - as before, the file path shouldn't be needed, only the deduced filename mask.
Lost files are probably being misreported (at least during initial phase), because the paths they walk do not match the physical representation on disk (due to striping and dir-per-part directories).

saxref  should:
- ensure runXRef is dealing with 1 plane a time (i.e. if multiple selected, process 1 at a time).
- store plane details (IPT) so accessible to other xref tasks during scan.
- during scanDirectories, if striped, detected when at stripe level, and 'skip'
- detect dir-per-part directory and 'skip'. But, check after recursing the dir, that contains no subdirs.
- Keep the filename mask only in cFileDesc
- Add scope/lfn to listOrphans. Build up in same way as 'basedir' is now.
Think we can get rid of baseDir altogether.
- during scanLogicalFiles, parse file paths fetched from parts (add helper func), to remove stripe and dir-part-part, so can marry the pathing with the cDirDesc tree.
NB: LATER- may be better to assume that the cDirDesc is a representation of scopes, and walk scopes of logical file (+part endpoints), instead of getting part directories.
- In listOrphans(cFileDesc), deduce file path from lfn, partNum and plane. Add a utility func that deduces and uses stripe num and dir-per-part if relevant (based on plane details).

@jackdelv

jackdelv commented on dali/sasha/saxref.cpp:
```diff
@@ -185,23 +213,49 @@ struct cFileDesc // no virtuals
 
     StringBuffer &getPartName(StringBuffer &buf,unsigned p)
     {
+        // In baremetal, buf can be prepoulated with replicate directory
```
Changed cFileDesc to contain just the file name.

jackdelv commented on dali/base/dautils.cpp:
```diff
@@ -3868,10 +3870,16 @@ extern da_decl void parseFileName(const char *name,StringBuffer &mname,unsigned
-                mname.append((d+1)-name, name).append(cur-(tailSlash+1), tailSlash+1);
+                if (dirs)
+                    dirs->append((d+1)-name, name);
+                mname.append(cur-(tailSlash+1), tailSlash+1);
```
Removed use of parseFileName from addfile. Now the striped and dir-per-part status are deduced from the directory structure while building up cDirDescs in scanDirectory and passed to recursive calls.

jackdelv commented on dali/base/dautils.cpp:
```diff
@@ -3834,7 +3835,8 @@ extern da_decl void parseFileName(const char *name,StringBuffer &mname,unsigned
-    if (mname.isEmpty())
+    // Assume that if prefix is passed in a match is required
+    if (prefix && prefix->isEmpty())
         throw makeStringExceptionV(-1, "Could not find matching prefix in plane definition for file %s", filename);
```
Instead of matching the prefix for each file, the XRefManager stores the current storage plane.

jackdelv commented:
@jakesmith I made the changes to remove adding path and scope to cFileDesc and instead pass the plane information and striped and dir-per-part statuses down to cFIleDesc::create. This avoids having to match the prefix and lookup storage planes for every file.

It seems in scanLogicalFiles there is more wrong than just not being able to include stripe and dir-per-part numbers. It is trying to match the current directory with the root. In bare-metal the root prefix looks like /var/lib/HPCCSystems/hpcc-data/thor based on which cluster is currently running xref. This will always fail to find files in directories for other components e.g. it will always say files under /var/lib/HPCCSystems/hpcc-data/roxie are potential lost files if doing the search on thor. In containerized this doesn't happen since the prefix won't contain the component name at the tail and will instead be /var/lib/HPCCSystems/hpcc-data. In this case the missing stripeNum and dir-per-part nums do cause an issue.

Other than that, my tests for found/orphan files are working and everything is showing up correctly in results on bm and containerized versions.

AttilaVamos commented:
@jackdelv There are some Unit test errors. See: https://github.com/hpcc-systems/HPCC-Platform/actions/runs/14222208906/job/39855851609?pr=19639
(I need to fix that test to report this kind of errors. I opened a JIRA to do that.)

jakesmith commented:
> @jackdelv There are some Unit test errors. See: https://github.com/hpcc-systems/HPCC-Platform/actions/runs/14222208906/job/39855851609?pr=19639 (I need to fix that test to report this kind of errors. I opened a JIRA to do that.)

@streeterd - hit similar problems when testing https://hpccsystems.atlassian.net/browse/HPCC-27051.
He has some pending changes to make to the k8s github action, which provide more insight when a deployment fails like this. See  https://hpccsystems.atlassian.net/issues/HPCC-33739

jackdelv commented:
@AttilaVamos @jakesmith I fixed the issue with the DFUXrefLibTests. I had made some changes to parseFileName and didn't notice the tests were failing. If the unittests are reporting errors, why is the github action being marked as "Succeeded"? Also, it seems like the DFUAccessTests failures are occurring in other PR's. Should I ignore them?

jackdelv commented:
@jakesmith I noticed another error is occurring that I have never seen before. 

> Waiting for deployment "sasha-xref" rollout to finish: 0 of 1 updated replicas are available...
error: timed out waiting for the condition

In the test: https://github.com/hpcc-systems/HPCC-Platform/actions/runs/14222208906/job/39856041969

This probably has something to do with how I instantiated the sasha xref service, but I'm not sure what I did wrong.

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -936,19 +972,23 @@ class CNewXRefManager: public CNewXRefManagerBase
-                    if (!scanDirectory(node,ep,path,drv,pdir,NULL))
+                if (isSpecialDir) {
+                    // Files inside a special directory should be scanned with the current directory as the parent
+                    if (!scanDirectory(node,ep,path,drv,pdir,NULL,!isStriped,isStriped))
```
I think isStriped shouldn't be a recursively passed property.
We know whether files should be in stripe directories or not from the outset (from the plane details).
scanDirectories should check that the 1st level is a stripe directory format, skip it like this if it is, report an error if not and ignore directory.

jakesmith replied:
>I was trying to use numDevices, but in bare-metal I have a test file that does not have a stripeDir despite numDevices>1 being in the plane info.

If a plane has numDevices and it is not a 'hostGroup' (!plane->hasProp("@hostGroup")) then it is striped.
@hostGroup planes have a list of planes, although supported in containerized, they are generally going to be BM planes mapped over a set of hosts.


>Also, if there is something in the plane details saying there should be stripeDir, does this apply to all files in the plane i.e. could there be a plane where some files are striped on it and some aren't?

No, if a plane implies striped, all files must be. It's because the plane defined the storage, the lfn's only know they layout based on the plane.
You could imagine a different scenario where the lfn's stored more detail - that might imply that the underlying original plane could change, and the existing lfn's remained associated with the old info, but that is not the design. Plane essentially need to be thought of as immutable, if different storage configuration were needed, then a new plane would need defining, and e.g. set as default, and pre-existing files would either remain on the old plane, or would need migrating.

jakesmith replied:
So:
1) isStriped should be deduced at the start of the scan (NB: dirPerPart could vary per file).
2) Should be remove as a member of cFileDesc
3) scanDirectory should assert that the top-level directory looks like a strip-directory, and reject the files if not.

NB: TBD (open a new sub JIRA place) - rejecting files both for this and other places, e.g. when "Directory Per Part %s contains other subdirectories.", should probably not short-circuit and ignore them, but add them, such that they are picked up as orphans, or misplaced.

jackdelv replied:
Removed isStriped from cFileDesc and added a check to scanDirectories to log an error when expected stripe directories are not found in the toplevel.

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -1235,21 +1318,19 @@ class CNewXRefManager: public CNewXRefManagerBase
+            StringBuffer path;
             if (drv)
-                setReplicateFilename(path,drv);
+                setReplicateFilename(addPathSepChar(path.append(basedir)),drv);
```
may be able to use constructPartFilename  vs getPhysicalPartName and not special case here.

jakesmith replied:
Why did lower casing causes issues? (see other comment).

jackdelv replied:
Removed lower casing in constructPartFilename because it was lower casing the prefix along with the filename.

jakesmith reviewed (changes requested):
@jackdelv - please see feedback.

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -997,7 +996,11 @@ class CNewXRefManager: public CNewXRefManagerBase
+                    // so that file parts with different physical locations can be correctly matched e.g.
+                    // /var/lib/HPCCSystems/hpcc-data/d1/somescope/otherscope/1/afile.1_of_2
+                    // /var/lib/HPCCSystems/hpcc-data/d2/somescope/otherscope/2/afile.2_of_2
+                    // These files would never be matched if we didn't build up the cDirDesc structure without the stripe and dir-per-part directories
```
It's more that we want the cDirDesc hierarchy to match the logical scope hierarchy, so that we can tie them together. We should be able to tell from the plane details whether the top-level directory under the prefix is expected to be a stripe directory.
At the physical level, the stripe dir and dir-per-part are implied/inserted based on meta details, i.e. you should be able to derive a physical path from a logical file scope + meta info. re. whether it's striped and/or dir-per-part.

jackdelv replied:
Updated comment.

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -1235,21 +1318,19 @@ class CNewXRefManager: public CNewXRefManagerBase
-                    f->getPartName(path,pn);
                     RemoteFilename rfn;
-                    rfn.setPath(grp->queryNode((pn+drv)%numnodes).endpoint(),path.str());
+                    rfn.setPath(grp->queryNode((pn+drv)%numnodes).endpoint(),getPhysicalPartName(path,storagePlane,lfn.get(), pn, f->N, f->isDirPerPart, f->isStriped).str());
```
>I'm not sure why it does this, but I didn't want to change since it is probably relied on for other uses.

I think it's because all logical files are caseless by definition.. e.g. a::b::c is same as A::B::C - physical files should be lower cases in DFS.

>which causes checkOrphanPhysicalFile to not find the file.

it failed because there was a physical file with mixed case? (there shouldn't be)

jackdelv replied:
Changed to use constructPartFilename and removed lower casing from the function.

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -936,19 +972,23 @@ class CNewXRefManager: public CNewXRefManagerBase
+                    if (!scanDirectory(node,ep,path,drv,pdir,NULL,!isStriped,isStriped))
                         return false;
+                    if (!isStriped && pdir->dirs.ordinality()>0)
+                        throw makeStringExceptionV(-1, LOGPFX "Directory Per Part %s contains other subdirectories.", path.str());
```
See other comment, we need to revisit these rejections, they should appear in the orphaned report I think (needs JIRA to be done after this one).

jackdelv replied:
Currently, it is calling OERRLOG to report that the dir per part contains sub dirs. Is there a better way to log what is happening?

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -936,19 +972,23 @@ class CNewXRefManager: public CNewXRefManagerBase
-                    // sure it is really a dirPerPart directory. Is an all numbers subdirectory valid in ecl?
-                    if (!scanDirectory(node,ep,path,drv,pdir,NULL))
+                if (isSpecialDir) {
+                    // Files inside a special directory should be scanned with the current directory as the parent
```
See my comment.

jakesmith reviewed (commented):
@jackdelv - changes look sensible. Please see follow up comments.

jackdelv commented on dali/sasha/saxref.cpp:
```diff
@@ -228,15 +230,23 @@ struct cFileDesc // no virtuals
-    StringBuffer &getName(StringBuffer &buf)
+    bool getName(StringBuffer &buf)
+    {
+        bool maskExists = filenameLen != 0;
```
Fixed.

jackdelv commented:
@jakesmith Removing the call to StringBuffer::lowercase in constructPartFilename provided the expected physical filename, so I have removed getPhysicalPartName. I removed isStriped from the parameters of scanDirectory and the cFileDesc member since it can be deduced from the plane info.

I've taken note of the two JIRAs you have mentioned. Am I missing anything?
- Revisit errors/exceptions being thrown to ensure XRef keeps running even if a file causes an error
- Reject Files not under stripe directory if plane details show directories are striped and subdirectories under dir-per-part directories.

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -903,6 +940,9 @@ class CNewXRefManager: public CNewXRefManagerBase
         }
         StringBuffer fname;
+        unsigned numDevices = 0;
+        if (pdir == root && !storagePlane->hasProp("@HostGroups"))
```
this should be `@hostGroup`.
Please check.

jackdelv replied:
Verified that `@hostGroup` is what is in the plane info. Fixed.

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -903,6 +940,9 @@ class CNewXRefManager: public CNewXRefManagerBase
         StringBuffer fname;
+        unsigned numDevices = 0;
+        if (pdir == root && !storagePlane->hasProp("@HostGroups"))
+            unsigned numDevices = storagePlane->getPropInt("@numDevices");
```
slightly surprised this didn't cause a compile error, unintended stack local 'unsigned'

jakesmith replied:
numDevices here and on line 943 should default to 1. Striped if >1

jakesmith replied:
you may have used 0 and >0, because constructPartFilename is adding a strip dir even if 1?
constructPartFilename should skip calcStripeNum if numDevices <= 1, and pass 0 into makePhysicalPartName

jakesmith replied:
also, it's going to be doing this for every directory as it recurses..
Better to move out, do once, and store result in a "isPlaneStriped" member for efficiency/clarity.

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -959,13 +960,16 @@ class CNewXRefManager: public CNewXRefManagerBase
                     if (!isdigit(*(dir++)))
                         isSpecialDir = false;
                 }
+                // If numDevices>0, check that top-level directories are striped
```
should be >1 (see earlier comments)

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -911,17 +951,43 @@ class CNewXRefManager: public CNewXRefManagerBase
+                bool isSpecialDir = true;
+                while (isSpecialDir && *dir) {
+                    if (!isdigit(*(dir++)))
+                        isSpecialDir = false;
```
shouldn't this set isStriped to false too?

jakesmith replied:
also, as it stands if just 'd' , isStriped will remain true when it should not (it would not be a isSpecialDir either).

jakesmith replied:
> also, as it stands if just 'd' , isStriped will remain true when it should not (it would not be a isSpecialDir either).

I think this is still true.. as the code stands, it won't enter the wild loop because if just "d", *dir will be \0. So will leave isDirStriped and isSpecialDir set to true incorrectly.

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -911,17 +951,43 @@ class CNewXRefManager: public CNewXRefManagerBase
+                        isSpecialDir = false;
+                }
+                // If numDevices>0, then pdir == root, and all directories should be stripe directories
+                if (numDevices>0 && (!isSpecialDir || !isStriped))
```
is pdir == root condition missing?

See other comment re. checking numDevices once, shouldn't redo per recursed directory.
To clarify distinction between the plane being striped and "isStriped" , can you rename "isDirStriped" here.

Aside from pdir==root, I think this condition is a bit wrong.
what if !isPlaneStriped && isDirStriped .. that should be an error too.

I think this expression needs to be something like:

if ((pdir==root) && (isPlaneStriped != isDirStriped)) // mismatch

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -911,17 +951,43 @@ class CNewXRefManager: public CNewXRefManagerBase
+                }
+                // If numDevices>0, then pdir == root, and all directories should be stripe directories
+                if (numDevices>0 && (!isSpecialDir || !isStriped))
+                    OERRLOG(LOGPFX "Top-level directory %s is not striped, but numDevices=%d", path.str(), numDevices);
```
can you verify what happens in this scenario..
e.g. if stripped and the top-level directory is not a strip directory.. do the files within it get reported as orphaned files, or.. ?

jackdelv replied:
Currently, if isPlaneStriped is true and isDirStriped is false, it will log the error `XREF: Top-level directory striping mismatch for /var/lib/HPCCSystems/hpcc-data/hthor: isPlaneStriped=1` and continue processing as normal. The mismatched directories aren't treated any differently than if the check was never there.

jakesmith reviewed (commented):
@jackdelv - please see comments.

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -712,8 +713,12 @@ class CNewXRefManager: public CNewXRefManagerBase
         if (plane)
+        {
             storagePlane.set(plane);
+            isPlaneStriped = !storagePlane->hasProp("@hostGroup") && (storagePlane->getPropInt("@numDevices", 1)>1);
```
trivial: no point in the default value in this context.

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -692,6 +692,7 @@ class CNewXRefManager: public CNewXRefManagerBase
     unsigned numuniqnodes = 0;
     Owned<IUserDescriptor> udesc;
     Linked<IPropertyTree> storagePlane;
+    bool isPlaneStriped;
```
trivial: even though itialized in ctor, it's good practice for all members (for new code at least) to have member initializers. Helps if code is ever refactored etc., and compiler is usually smart enough not to initialize twice anyway.

jakesmith reviewed (commented):
@jackdelv - 1 issue re. "d" still I think + 2 trivial comments.

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -188,15 +192,24 @@ struct cFileDesc // no virtuals
     }
 
-    StringBuffer &getName(StringBuffer &buf)
+    bool getName(StringBuffer &buf)
```
trivial/not new: should be a const method really (as should get some of the members below)

jackdelv commented:
@jakesmith Fixed the trivial items. I added a check to the initializer for isDirStriped that dir[1] != '\0'. It seemed to be the best way to fix it, but is there a better way? Can the whole stripe/dirperpart check be refactored in a cleaner way?

jackdelv commented:
@jakesmith Opened two JIRAs relating to our discussion:

https://hpccsystems.atlassian.net/browse/HPCC-33889
https://hpccsystems.atlassian.net/browse/HPCC-33890

jakesmith reviewed (commented):
> @jakesmith Fixed the trivial items. I added a check to the initializer for isDirStriped that dir[1] != '\0'. It seemed to be the best way to fix it, but is there a better way? Can the whole stripe/dirperpart check be refactored in a cleaner way?

I think it works as is.
I'd probably have introduced a isRestDigits inline helper and code like:
```
                bool isDirStriped = false; // format 'd[0-9]+'
                bool isSpecialDir = false; // isDirStriped || dirPerPart format '[0-9]+'
                if (*dir == 'd')
                {
                    dir++;
                    isDirStriped = (*dir != '\0') && isRestDigits(dir);
                    isSpecialDir = isDirStriped;
                }
                else
                    isSpecialDir = isRestDigits(dir);
```

untested!

@jackdelv - looks good I think. Please squash, and I'll hopefully do a final review in the morning.

jackdelv committed:
HPCC-30365 Add XREF Sasha service to K8s

jackdelv commented on dali/sasha/saxref.cpp:
```diff
@@ -911,17 +952,45 @@ class CNewXRefManager: public CNewXRefManagerBase
+                while (isSpecialDir && *dir) {
+                    if (!isdigit(*(dir++))) {
+                        isSpecialDir = false;
+                        isDirStriped = false;
```
Fixed.

jakesmith reviewed (commented):
@jackdelv - a few final comments.

jackdelv commented:
@jakesmith Added commits for your comments. Back to you.

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -2066,6 +2066,7 @@ int main(int argc, char* argv[])
 #endif
 
 
+constexpr float maxMemPercentage = 0.9; // In containerized, leave some headroom for the pod
```
nitpicking: should be static too.

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -716,7 +717,8 @@ class CNewXRefManager: public CNewXRefManagerBase
         {
             storagePlane.set(plane);
-            isPlaneStriped = !storagePlane->hasProp("@hostGroup") && (storagePlane->getPropInt("@numDevices")>1);
+            numDevices = storagePlane->getPropInt("@numDevices", 1);
```
need to be careful where used, e.g. on line 1326.. it is being used to indicate striping too, which will be wrong if `@hostGroup`.
I think you rename this to numStripedDevices - and set:
```
numStripedDevices = isPlaneStriped ? storagePlane->getPropInt("@numDevices", 1) : 1
```

jakesmith reviewed (commented):
@jackdelv - left a couple more comments. Please rebase to master.

jackdelv commented:
@jakesmith Rebased to master and added minor fixes from your comments. Back to you.

jackdelv commented:
@jakesmith There was an issue where the XRef pod was crashing saying no access to dali. I believe it was due to a change to saserver.cpp:main that adds an explicit check for access=dali. I have added a fix for that issue.

jakesmith commented:
> @jakesmith There was an issue where the XRef pod was crashing saying no access to dali. I believe it was due to a change to saserver.cpp:main that adds an explicit check for access=dali. I have added a fix for that issue.

That makes sense, and in fact I believe the github k8s test was failing, because with the "dali" for xref in this change, the xref sasha pod would have no egress access (k8s NetworkPolicy) to access it, and was unable to connect as a dali client.

jakesmith reviewed (commented):
@jackdelv - looks good. Please squash, and I'll give it one more pass hopefully.

jackdelv commented:
@jakesmith Squashed.

jakesmith commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -757,6 +769,16 @@ struct CLogicalNameEntry: public CInterface
+        if (!plane)
+            throw makeStringExceptionV(0, "Plane definition \"%s\" is missing for File", grpname.str());
+        unsigned numDevices = plane->numDevices();
+        bool r = replicate?1:0;
```
should be unsigned r =

jakesmith reviewed (commented):
@jackdelv - looks good.

As discussed, the dfuxreflib should likely be deprecated altogether. It is not the default or ever used in BM afaik (saxref is the standard setuip).
It would be good to assert that it is never used in k8s, and remove changes to it in this PR.

jackdelv commented:
@jakesmith Removed the dfuXRefLib changes. Since the constructPartFilename function was updated and is used in saxref, I changed dfuXRefLib to use the now deprecated version. I also added an assert that containerized should be using the new saxref path.

jakesmith commented on esp/services/ws_dfu/ws_dfuXRefService.hpp:
```diff
@@ -101,7 +101,13 @@ class CXRefExBuilderThread : public Thread
+                {
+                    // MORE: Containerized should only be using saxref and dfuXRefLib should probably be removed
+                    assert(xRefNode->useSasha());
+                    xRefNode->setStatus("Submitted");
```
I think better to remove this line and the else. i.e. just the assert (should use assertex),

jakesmith replied:
Good to catch/assert here, but what I meant/was referring to, was the code in saxref.cpp:
```
        bool eclwatchProviderDefault = false;
        // NB: Default to start Xref using sasha service in containerized without having to change values.yaml
        // eclwatchProvider sets useSasha in call to setSubmittedOk
        if (isContainerized())
            eclwatchProviderDefault = true;
        bool eclwatchprovider = props->getPropBool("@eclwatchProvider",eclwatchProviderDefault);
```

That should explicitly only support the eclwatch provide option in containerized, e.g.:
```
        bool eclwatchprovider = true;
        if (!isContainerized()) // NB: containerized does not support xref any other way.
        {
            // eclwatchProvider sets useSasha in call to setSubmittedOk
            eclwatchprovider = props->getPropBool("@eclwatchProvider");
        }
```

jakesmith reviewed (commented):
@jackdelv - a couple of comments.

I notice some whitespaces creeping in (at end of lines).
(more obvious when using git diff on the CLI)
I haven't used it personally, but there is a vscode option:
```
"files.trimTailingWhitespace" : true
```
which may be useful

jackdelv commented:
@jakesmith Fixed the assert. I also turned on that setting to hopefully prevent that from happening in the future. Which files had trailing whitespaces? I know the dfuXRefLib diff shows whitespace changes still, but those are removals.

jakesmith commented:
> @jakesmith Fixed the assert. I also turned on that setting to hopefully prevent that from happening in the future. Which files had trailing whitespaces? I know the dfuXRefLib diff shows whitespace changes still, but those are removals.

hm, I don't see them now. Maybe I was looking at wrong diff.

jakesmith reviewed (commented):
@jackdelv - looks good. Please squash.

jakesmith commented:
@jackdelv - now that candidate-9.12.x exists, this needs to be rebased onto it.

jackdelv commented:
@jakesmith Squashed, and Rebased on 9.12.x. Back to you.

jakesmith reviewed (approved):
@jackdelv - looks good.

@ghalliday - please take a look.

ghalliday commented on dali/sasha/saxref.cpp:
```diff
@@ -123,7 +125,7 @@ struct cFileDesc // no virtuals
-    static cFileDesc * create(CLargeMemoryAllocator &mem,const char *_name,unsigned n)
+    static cFileDesc * create(CLargeMemoryAllocator &mem,const char *_name,unsigned n,bool d,unsigned fnLen)
     {
         size32_t sl = strlen(_name); 
```
future: add an assert that fnLen <= 255

ghalliday commented on dali/sasha/saxref.cpp:
```diff
@@ -1082,7 +1157,7 @@ class CNewXRefManager: public CNewXRefManagerBase
                                 if (dir.length()&&isPathSepChar(dir.charAt(dir.length()-1)))
                                     dir.setLength(dir.length()-1);
-                                unsigned drv = getPathDrive(dir.str()); // should match c
+                                unsigned drv = isContainerized() ? getPathDrive(dir.str()) : 0; // should match c
```
Is this test around the correct way?

jackdelv replied:
No, that is wrong. It should be `isContainerized() ? 0 : getPathDrive(dir.str());`

ghalliday reviewed (commented):
@jackdelv looks good.  One question, otherwise I am happy to merge.  (One additional minor comment.)

ghalliday reviewed (approved):
@jackdelv please squash

jackdelv commented:
@ghalliday Squashed.

ghalliday commented on dali/sasha/saxref.cpp:
```diff
@@ -2086,31 +2169,58 @@ class CSashaXRefServer: public ISashaServer, public Thread
+        ForEachItemIn(i1, list) {
+            const char *planeName = list.item(i1);
+            Owned<IPropertyTree> plane = getStoragePlane(planeName);
+            if (isContainerized) {
```
isContainerized()?
